src/client/Microsoft.Identity.Client/PoP/PoPAuthenticationScheme.cs

```diff
@@ -0,0 +1,132 @@
+        private readonly byte[] _keyThumbprint;
+
+        /// <summary>
+        /// Creates POP tokens, i.e. tokens that are bound to an HTTP request and are digitally signed.
+        /// </summary>
+        /// <remarks>
+        /// Currently the signing credential algorithm is hard-coded to RSA with SHA256. Extensibility should be done
+        /// by integrating Wilson's SigningCredentials
+        /// </remarks>
+        public PoPAuthenticationScheme(Uri requestUri, HttpMethod httpMethod, IPoPCryptoProvider popCryptoProvider)
+        {
+            _httpMethod = httpMethod ?? throw new ArgumentNullException(nameof(httpMethod));
+            _requestUri = requestUri ?? throw new ArgumentNullException(nameof(requestUri));
+            _popCryptoProvider = popCryptoProvider ?? throw new ArgumentNullException(nameof(popCryptoProvider));
+
+            _keyThumbprint = ComputeRsaThumbprint(_popCryptoProvider.CannonicalPublicKeyJwk);
+            KeyId = Base64UrlHelpers.Encode(_keyThumbprint);
+        }
+
+        public string AuthorizationHeaderPrefix => PoPRequestParameters.PoPAuthHeaderPrefix;
+
+        /// <summary>
+        /// For PoP, we chose to use the base64(jwk_thumbprint)
+        /// </summary>
+        public string KeyId { get; }
+
+        public IDictionary<string, string> GetTokenRequestParams()
+        {
+            return new Dictionary<string, string>() {
+                { OAuth2Parameter.TokenType, PoPRequestParameters.PoPTokenType},
+                { PoPRequestParameters.RequestConfirmation, ComputeReqCnf()}
+            };
+        }
+
+        public string FormatAccessToken(MsalAccessTokenCacheItem atItem)
+        {
+            var header = new JObject
+            {
+                { JsonWebTokenConstants.ReservedHeaderParameters.Algorithm, JsonWebTokenConstants.Algorithms.RsaSha256 },
+                { JsonWebTokenConstants.ReservedHeaderParameters.KeyId, KeyId },
+                { JsonWebTokenConstants.ReservedHeaderParameters.Type, PoPRequestParameters.PoPTokenType}
+            };
+
+            JToken popAssertion = JToken.Parse(_popCryptoProvider.CannonicalPublicKeyJwk);
+
+            var payload = new JObject(
+                new JProperty(PoPClaimTypes.At, atItem.Secret),
+                new JProperty(PoPClaimTypes.HttpMethod, _httpMethod.ToString()),
+                new JProperty(PoPClaimTypes.Host, _requestUri.Host),
+                new JProperty(PoPClaimTypes.Path, _requestUri.AbsolutePath),
+                new JProperty(PoPClaimTypes.Nonce, CreateNonce()),
+                new JProperty(PoPClaimTypes.Cnf, new JObject(
+                    new JProperty(PoPClaimTypes.JWK, popAssertion))));
+
+            return CreateJWS(payload.ToString(Json.Formatting.None), header.ToString(Json.Formatting.None));
+        }
+
+        private static string CreateNonce()
+        {
+            // Guid with no hyphens
+#if NETSTANDARD || WINDOWS_APP
+            return Guid.NewGuid().ToString("N");
```

Comment:
>Guid.NewGuid().ToString("N") [](start = 19, length = 28)

how is this different from doing Guid.NewGuid()?

Reply:
Guid.NewGuid() returns an object of type `Guid` so we need to tranform it into a string. .NET can do that for us because each and every object implements a method called `ToString`, although note that in most cases you get a string represenation like `this is an array` instead of `1 2 3`. In case of GUIDs, .net folks have implemented ToString for us and it returns a string in the format `xxxx-xxxx-xxxxxxxxx-xxxx`. But I do not want hyphens on there (this is what I discussed with George), and the ToString("N") takes care of that.

Reply:
cool. thanks for the explanation.

---
In reply to: [352707770](https://github.com/AzureAD/microsoft-authentication-library-for-dotnet/pull/1525#discussion_r352707770) [](ancestors = 352707770)